### ISSUE TITLE
transformer ASR overfitting on librispeech100h

### ISSUE THREAD

**Issue by BuaaAlban** [espnet/espnet#3761]
**Describe the bug**
Not a bug, ask for help and training techniques.
The baseline config is the config for training transformer ASR on librispeech **960h**, and I have trained model on 960h sucessfully. This config also **works** for **aishell-1(160h),** but I can't get results on librispeech**100h**.

**Basic environments:**
 - OS information: Ubuntu 18.04.1 LTS
 - python version: [e.g. 3.8.5 (default, Sep 24 2020, 16:55:52)  [GCC 7.5.0]`]
 - espnet version: [e.g. espnet 0.9.6]
 - Git hash [e.g. c84da5743b7ef70c0c6212715859bdebdcf873b2]
   - Commit date [e.g. Tue Sep 1 09:32:54 2020 -0400]
 - pytorch version [e.g. pytorch 1.7.1]
 

**Environments from `torch.utils.collect_env`:**
Collecting environment information...
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.1 LTS
GCC version: (GCC) 7.5.0
CMake version: version 3.10.2

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: Tesla V100-PCIE
GPU 1: Tesla V100-PCIE
GPU 2: Tesla V100-PCIE
GPU 3: Tesla V100-PCIE

Nvidia driver version: 470.63.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
Versions of relevant libraries:
[pip3] numpy==1.20.1
[pip3] pytorch-wpe==0.0.1
[pip3] torch==1.7.1
[pip3] torch-complex==0.2.1
[conda] Could not collect

**Task information:**
 - Task: ASR
 - librispeech 100h
 - ESPnet2

**Experience and logs*
The followings are the configs I  have tried. 
[train_csasr_transformer.txt](https://github.com/espnet/espnet/files/7495766/train_csasr_transformer.txt)
![loss](https://user-images.githubusercontent.com/25258181/140718285-b277db89-50bf-41a2-b261-b60076d5f143.png)
![lr_0](https://user-images.githubusercontent.com/25258181/140718317-07c3190d-02e4-467c-84dc-a712c3bd2012.png)

It overfits and I tried to increase dropout rate, though there is no overfitting, the training process doesn't converge.
[train_csasr_transformer_dropout.txt](https://github.com/espnet/espnet/files/7495767/train_csasr_transformer_dropout.txt)

![loss](https://user-images.githubusercontent.com/25258181/140719718-7cb1a93c-9800-44f7-b8f1-9a320c529207.png)

I found the warm up step is not right for the 100h dataset, which may case the problem(lr is too small) so I decrease the warmup step to 400

[train_csasr_transformer_100h_2000warmup.txt](https://github.com/espnet/espnet/files/7495764/train_csasr_transformer_100h_2000warmup.txt)

[train_csasr_transformer_100h_400warmup.txt](https://github.com/espnet/espnet/files/7495762/train_csasr_transformer_100h_400warmup.txt)
![loss](https://user-images.githubusercontent.com/25258181/140717976-41a48d83-5a79-437b-810c-dcf57e5fe599.png)
![lr_0](https://user-images.githubusercontent.com/25258181/140717995-c8d53360-4d26-427f-9bd2-aa204c058dbe.png)

It still overfit and I have no idea what to do next.
*
Do you have any suggestions ？ Thanks a lot if you could help
I rying other dropout settings because the dropout I have augmented  seems too big

**Comment by sw005320** [comment]
Did you reduce `lr`?

**Comment by BuaaAlban** [comment]
> Did you reduce `lr`?

It appears to be overfited, so I didn't try reducing 'lr'. 
I will try to reduce it in the original setting except for 'lr 0.002 --> 0.0002 and  warmup steps 25000 -> 400'

**Comment by pyf98** [comment]
I think there are a few things you could try.

1. Keep the dropout probability as the default value 0.1. We usually don't need to change this hyperparameter. A too large value can make the model diverge.
2. Try different learning rates and warmup steps. For example, you can try `lr=5e-4, 1e-3, 2e-3, 5e-3, 1e-2, etc`. I think `1e-3` or `2e-3` could work. The number of warmup steps is more tricky. It depends on other hyperparameters: `accum`, `batchbins`. It is also influenced by the size of the dataset. In your last figure, the max lr is achieved after 6 epochs. Maybe you can make it longer, such as 15 epochs.

**Comment by BuaaAlban** [comment]
With dropout=0.1, I tried all the lr with warmup steps=1000 (around 15 epochs) and they all didn't work. 
Tuning is really a trick work.

**Comment by sw005320** [comment]
Did you keep warmup steps=25000 for this search?
warmup steps=1000 makes your learning rate too steep.
I feel that you seem to change too many hyperparameters.
Before changing various hyperparameters at the same time, I recommend you just start from our default transformer configs in the repo and only change the learning rate.

**Comment by BuaaAlban** [comment]
Actually I didn't keep warmup steps=25000 . Due to the dataset size, lr will keep going up during training with warmup=25000. Though it works with aishell with the arugmenting lr.
![image](https://user-images.githubusercontent.com/25258181/143366643-89b3436e-40bf-4f3c-a871-103bbba4a78d.png)

warup steps = 1000 seems better
![image](https://user-images.githubusercontent.com/25258181/143366939-ad057383-bfa4-4046-bd9d-df01b45af040.png)

And for the lr rate exp recommended by @pyf98 , I only change the lr  and warmup steps=1000.  All other parameters are the same as that one which sucessfully trained on librispeech960h and aishell. 
Will the bpe size have the effects? I set it to 2k for librispeech100h

**Comment by pyf98** [comment]
Not very sure if the BPE size plays a key role here. I actually used 300 and it worked. But I was using a different model. Maybe you could try a smaller BPE size, e.g. 300. @BuaaAlban

**Comment by BuaaAlban** [comment]
Thanks， I will try it.
